File: day3/part2.py
def readlines(filename):
    with open(filename, "r") as f:
        return [line.strip() for line in f.readlines()]
    


def findLargestBatteryIndex(bank):
    maxIndex = 0
    maxValue = -1
    for i in range(len(bank)):
        value = int(bank[i])
        if value > maxValue:
            maxValue = value
            maxIndex = i
    return maxIndex


def findLargestGroup(bank, size):
    if size == 1:
        return bank[findLargestBatteryIndex(bank)]
    elif size == len(bank):
        return bank
    else:
        nextIndex = findLargestBatteryIndex(bank[0:len(bank)+1-size] )
        nextBattery = bank[nextIndex]
        nextGroup = findLargestGroup(bank[nextIndex+1:], size-1)
        largestGroup =  nextBattery + nextGroup
        print(f"Largest group of size {size} in bank {bank} is {largestGroup}")
        return largestGroup


lines = readlines("input.txt")
# total = sum(findMaxGroup(line, 12) for line in lines)
total = sum(int(findLargestGroup(line, 12)) for line in lines)
print(total);

File: day3/part2.js
const fs = require('fs');

function readlines(filename) {
  return fs.readFileSync(filename, 'utf-8')
    .split('\n')
    .map(line => line.trim())
    .filter(line => line.length > 0);
}

function findLargestBatteryIndex(bank) {
  let maxIndex = 0;
  let maxValue = -1;
  for (let i = 0; i < bank.length; i++) {
    const value = parseInt(bank[i]);
    if (value > maxValue) {
      maxValue = value;
      maxIndex = i;
    }
  }
  return maxIndex;
}

function findLargestGroup(bank, size) {
  if (size === 1) {
    return bank[findLargestBatteryIndex(bank)];
  } else if (size === bank.length) {
    return bank;
  } else {
    const nextIndex = findLargestBatteryIndex(bank.slice(0, bank.length + 1 - size));
    const nextBattery = bank[nextIndex];
    const nextGroup = findLargestGroup(bank.slice(nextIndex + 1), size - 1);
    const largestGroup = nextBattery + nextGroup;
    console.log(`Largest group of size ${size} in bank ${bank} is ${largestGroup}`);
    return largestGroup;
  }
}

const lines = readlines('input.txt');
const total = lines.reduce((sum, line) => sum + parseInt(findLargestGroup(line, 12)), 0);
console.log(total);